Add option to drop partial buckets from date_histogram visuals

blfrantz opened the pull request:
Closes [#2806](https://github.com/elastic/kibana/issues/2806).

Note: description has changed since the original version to reflect a new implementation.

Date Histograms often begin and/or end with incomplete buckets, which makes the beginning/end of time-series charts appear to show steep up/down trends which can be misleading or alarming.  This feature adds a new "Drop partial buckets" option for the Date Histogram aggregation.  It only appears/applies when the chosen field is the same as the index's Time Filter field (that's the only case where this feature makes sense).

![drop_bucket](https://user-images.githubusercontent.com/25988482/42137455-11689afc-7d32-11e8-8be7-c7d52b9d94ae.JPG)

When selected, any buckets which span more time than is covered by the query's Time Range will be removed from the chart.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

ppisljar commented:
thanks a lot for your contribution @blfrantz !

As you mention this is something aggregation specific. Adding it to handler.js means it will only work with pie, area, line, bar, heatmap and gauge charts. It won't with all the others (maps, metric, tagcloud, more to come, any 3rd party visualizations).

I would suggest implementing this as an aggregation option. The date histogram aggregation config and logic is in `ui/agg_types/buckets/date_histogram.js`. This way this option can be completely independent from the visualization. 

The actual conversion of data happens in `tabify` response handler, this is probably where you should check for this option and skip partial buckets. `ui/vis/response_handlers/tabify.js` and `basic.js` in same folder use the tabify functions defined in `ui/agg_response/tabify`.

let me know what you think and if this is helpful.

blfrantz commented:
@ppisljar, any thoughts?  Thanks!

ppisljar commented:
sorry for late reply.

in your response handler you can access the timefilter on `vis.filters.timeRange`

preferably you would be able to figure this out of elastic search response, but i am not sure if that is possible.

blfrantz commented:
No problem, @ppisljar .

So in the relative search case, timeRange only contains something like `{from: "now-15m", to: "now", mode: "quick"}`.  Without knowing when the query fired, that's not really useful (and even if I did, it'd be ugly to calculate).  As for the ES response, the data returned by the _msearch query doesn't appear to have any timestamp info aside from the bucket times, but those don't necessarily align with the queried range (if they did, this whole feature would be unnecessary).

Unless this information is available somewhere I'm not aware of (or not that hard to pass in), I'm beginning to think my original approach may be more realistic.  In the same way that the "partial data" tooltip only shows up for some types of visuals (not all where it could be relevant), I'm wondering if having this feature exist only for some of the relevant cases (arguably the ones that matter most) is enough for now?  What do you think?  If you're ok with that approach, can you provide feedback on my 4th question (about test frameworks)? 

Thanks!

ppisljar commented:
```
import { getTime } from 'ui/timefilter/get_time';

const parsedTime = getTime(vis.indexPattern, vis.filters.timeRange);
```

let me know if this helps

blfrantz commented:
@ppisljar thanks for that tip, it worked great.  I've updated the PR with the new implementation, let me know what you think.

ppisljar commented on src/ui/public/agg_response/tabify/_buckets.js:
```diff
@@ -81,4 +82,20 @@ TabifyBuckets.prototype._orderBucketsAccordingToParams = function (params) {
 };
 
+TabifyBuckets.prototype._dropPartials = function (params, timeRange) {
+  if (params.drop_partials && !this.objectMode && this.buckets.length > 1) {
```
rather than wrapping the whole thing in an `if` statement return early:

`if (!params.drop_partials || this.objectMode || this.buckets.length <= 1) { return; }`

ppisljar replied:
maybe move the `if (params.drop_partials)` out of this function ?

also join with below if statement ... and extract to variables so its easier to read:
```
 const isTimeField = params.field.name === timeRange.name;
if (this.objectMode || this.buckets.length <= 1 || !isTimeField) { 
 return;
}
```

blfrantz commented:
Thanks @ppisljar.  That makes sense, and does make the code organization more elegant/intuitive.  However I'm finding that it makes the dropPartials function itself rather more complicated and therefore more brittle.

My current approach handles multiple series on a single chart, as well as split charts, as-is.  However, because of the way the tabified data is structured, things that were simple become a bit complicated.  

In the no-split-charts case, the following works (called with `dropPartialRows(write.response(), write.timeRange)` toward the end of `tabifyAggResponse()` in `tabify.js`:
```
function dropPartialRows(tableGroup, timeRange) {
  tableGroup.tables.forEach(table => {
    if (table.rows.length < 2) return;

    const aggIndex = _.findIndex(table.columns, {
      'aggConfig': {
        '_opts': {
          'type': 'date_histogram',
          'params': {
            'field': timeRange.name,
            'drop_partials': true
          }
        }
      }
    });

    if (aggIndex !== -1) {
      const time0 = table.rows[0][aggIndex].key;
      const time1 = _.find(table.rows, r => {
        if (r[aggIndex].key !== time0) return true;
        else return false;
      })[aggIndex].key;

      const interval = time1 - time0;

      table.rows = table.rows.filter(arr => {
        if (arr[aggIndex].key < timeRange.gte) return false;
        if (arr[aggIndex].key + interval > timeRange.lte) return false;
        return true;
      });
    }
  });
}
```

That's not too bad, but some things that could be simple (like determining the interval) get a bit complicated because tabify's rows array contains a separate row for every value on each series, which means there can be multiple rows per time bucket, which means I have to scan the array until I find a new time to determine the interval.  Previously, I could just compare the first and second items in the array.

Furthermore, this code doesn't currently handle the split charts case, because in that event tableGroups contains tables of tables, and so I'd need to process this recursively or something.  That's not a big deal, but as this gets more complex (and computationally expensive) I wanted to see if you thought this was still the best approach.  Apologies if I'm missing something.  Thanks!

timroes commented on src/ui/public/vis/response_handlers/basic.js:
```diff
@@ -74,7 +75,8 @@ const BasicResponseHandlerProvider = function (Private) {
           asAggConfigResults: true,
-          isHierarchical: vis.isHierarchical()
+          isHierarchical: vis.isHierarchical(),
+          timeRange: getTime(vis.indexPattern, vis.filters.timeRange).range
```
@ppisljar Since we talked earlier about removing `vis.filters`, should we introduce a new reference here? Do we have an idea how we remove that in the future?

ppisljar replied:
`vis.filters` is going away in the long run, but the response handlers will still get the access to the `timeRange` (probably being passed in directly as a parameter in the future) so this should not be a problem.

markov00 replied:
@blfrantz If you create a visualization without a time field you will get an error for this line because the `getTime` will return `undefined` if not time field is specified on the selected `indexPattern`.

You can verify the error using the `shakespeare` dataset https://www.elastic.co/guide/en/kibana/current/tutorial-load-dataset.html and just creating a barchart with that index.

blfrantz commented:
@ppisljar Thanks, I've cleaned up the conditionals for _dropPartials, let me know how it looks now.  I still need to rebase/retest before this is ready to pull but waiting to see if any further tweaks are needed per @timroes comments.

gchaps commented on src/ui/public/agg_types/controls/drop_partials.html:
```diff
@@ -0,0 +1,11 @@
+    &nbsp;
+    <icon-tip
+      position="'right'"
+      content="'Removes buckets that include times not covered by the Time Range.'"
```
@timroes 

Maybe something like this:

Remove buckets that span time outside the time range so the histogram doesn't start and end with incomplete buckets.

blfrantz commented:
@ppisljar & @timroes: I just force-pushed a cleanup of the commit history to remove all the old iterations.  As part of this I added the new tooltip text recommended by @gchaps.  I also just rebased onto the latest upstream, fixed a couple of my tests that were broken by some recent upstream changes, and sanity tested that the feature still works as expected.  Awaiting any further instructions.  Thanks!

blfrantz commented:
@ppisljar & @timroes: I know you guys are busy, but just wanted to make sure this was still on your radar.  Thanks!

timroes commented:
Jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7569/)

blfrantz commented:
Do I need to do anything about the build failure?  The result seems to have been purged, but when I looked at it when it was still available the test failure looked like an intermittent test timeout, not a problem with my changes.

timroes commented:
We had a couple of test stabilizations in the past week. Could you perhaps merge master into your PR, so we have tests run more stable on that PR too?

blfrantz commented:
@timroes Sure, done.

blfrantz commented:
@timroes Any update on this?

blfrantz commented:
@ppisljar @timroes  Hey guys, this PR still something you're interested in?  Can someone kick off another test run (I'm assuming I can't)?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/503/)

ppisljar commented:
jenkins, test this

ppisljar commented:
@blfrantz i am really sorry for the delays on this. I still think this is a great addition. Tim is out for vacation this week @markov00 can you take a look at this ?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/516/)

markov00 commented:
jenkins test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/526/)

markov00 commented on src/ui/public/vis/response_handlers/tabify.js:
```diff
@@ -32,7 +33,8 @@ const TabifyResponseHandlerProvider = function (Private) {
           asAggConfigResults: _.get(vis, 'type.responseHandlerConfig.asAggConfigResults', false),
-          isHierarchical: vis.isHierarchical()
+          isHierarchical: vis.isHierarchical(),
+          timeRange: getTime(vis.indexPattern, vis.filters.timeRange).range
```
I think the same will also happen here if you don't have a time field in your index pattern

markov00 commented on src/ui/public/agg_response/tabify/_buckets.js:
```diff
@@ -81,4 +84,23 @@ TabifyBuckets.prototype._orderBucketsAccordingToParams = function (params) {
+
+  this.buckets = this.buckets.filter(bucket => {
+    if (bucket.key < timeRange.gte) return false;
+    if (bucket.key + interval > timeRange.lte) return false;
```
nit: it's better to align with our styleguide [always-use-braces-for-conditionals-and-loops](https://github.com/elastic/kibana/blob/master/style_guides/js_style_guide.md#always-use-braces-for-conditionals-and-loops)

markov00 commented on src/ui/public/agg_response/tabify/_buckets.js:
```diff
@@ -81,4 +84,23 @@ TabifyBuckets.prototype._orderBucketsAccordingToParams = function (params) {
+  if (!timeRange ||
+    this.buckets.length <= 1 ||
+    this.objectMode ||
+    params.field.name !== timeRange.name) return;
```
nit: it's better to align with our styleguide always-use-braces-for-conditionals-and-loops

markov00 commented on src/ui/public/agg_response/tabify/_buckets.js:
```diff
@@ -38,7 +38,10 @@ function TabifyBuckets(aggResp, aggParams) {
-  if (this.length && aggParams) this._orderBucketsAccordingToParams(aggParams);
+  if (this.length && aggParams) {
+    this._orderBucketsAccordingToParams(aggParams);
+    if (aggParams.drop_partials) this._dropPartials(aggParams, timeRange);
```
nit: it's better to align with our styleguide always-use-braces-for-conditionals-and-loops

blfrantz commented:
@markov00 Thanks for the feedback and tip about that test failure.  I've merged and made the requested fixes.  I also confirmed that the failing Shakespeare test passes locally now.

markov00 commented:
jenkins test this

markov00 commented:
jenkins test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/598/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/607/)

markov00 commented:
jenkins test this, want to be sure it was a flaky  CI test failure

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/617/)

blfrantz commented:
What's the next step here?

timroes commented:
Jenkins, test this - I'll give it a last test run, then will merge it.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1299/)

timroes commented:
Hi Brian,

thanks a lot for your first contribution to Kibana! I just merged your commit on master and it will be backported to 6.x, so that these changes will be released in 6.5.

Cheers
Tim

blfrantz commented:
Awesome, thanks everyone!

LeeDr commented:
@timroes or @markov00 or @ppisljar can any of you help me understand why I don't see this option here on this 6.5.0 visualization?  This one is using ecommerce sample data, and I also tried a new line chart from makelogs data. 

![image](https://user-images.githubusercontent.com/13542669/48294623-0d341f80-e44b-11e8-8f0d-a5f8ebbb75de.png)

ppisljar commented:
@LeeDr https://github.com/elastic/kibana/pull/25520

elasticmachine commented:
Pinging @elastic/kibana-app

Benny-Git commented:
We just updated to 6.5.2, and I still don't see the checkbox option, even for a newly created visualization.
According to [the release notes](https://www.elastic.co/guide/en/kibana/6.5/release-notes-6.5.2.html), #25520 should be included in 6.5.2?

timroes commented:
@Benny-Git unfortunately that PR was missing the backport to 6.5 :disappointed: This should be released in a future 6.5 patch release instead, sorry.

@gchaps Could you perhaps remove `Fixes option for showing partial buckets #25520` from the 6.5.2 release notes?

Benny-Git commented:
@timroes thanks for the very quick response.
I'll just manually update my few relevant saved objects with `"drop_partials": true` for now :)

gchaps commented:
@timroes Done.
